### PR TITLE
Added testcase :Share and remove backup location

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #
https://portworx.atlassian.net/browse/PA-330
testcase : Share and remove backup location
Added method ShareBackupWithMultipleUsers
**Special notes for your reviewer**:
Aetos link: http://aetos.pwx.purestorage.com/resultSet/testSetID/81099
Latest aetos link after all the changes: http://aetos.pwx.purestorage.com/resultSet/testSetID/81608

### CODE REVIEW

File: tests/backup/backup_test.go
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2929]: Add long.InfoD here to report to dashboard

File: tests/backup/backup_test.go
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2935]: same here

File: tests/backup/backup_test.go
Comment (lsrinivas-pure) [inline tests/backup/backup_test.go:2940]: same here

**Review by lsrinivas-pure — CHANGES_REQUESTED** [review]

File: tests/backup/backup_test.go
Comment (mkoppal-px) [inline tests/backup/backup_test.go:5572]: I think a better name for this function would be `ShareBackupWithUsersAndAccessAssignment`
Let me know what do you think.

File: tests/backup/backup_test.go
Comment (mkoppal-px) [inline tests/backup/backup_test.go:5604]: Why is this in upper case? Do you want to `AccessUserBackupContext` to be used outside the package?

**Review by mkoppal-px — CHANGES_REQUESTED** [review]
I think your test is not as per the steps in testrail. Testrail steps are as follows:

```
1. Take a set of backups and share it with multiple users.
2. Remove the backup location after sharing it with all users.
3. Add it back to the cluster and check from other users if they show up.
4. Reshare backups and see if they are successful or not.
```

But in your automated test steps 3 and 4 are missing. You are taking the backup again which is not needed.
We have to delete the backup location and add it again. Then re-share it with all users and check the access levels.

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:2945]: Can we add return to `CreateBackupLocation()`, let's be consistent across similar functions, I see that `DeleteBackupLocation()` returns error, I don't see a reason why `CreateBackupLocation()` should be written in same style

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:5601]: Can we not define err and then return it, can we just return like
```
if err != nil {
			return AccessUserBackupContext, fmt.Errorf("unable to share backup %s with user %s Error: %v", backupNames[i], user, err)
}
```

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:2929]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:2940]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:5572]: sure.Done!!

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:5604]: changed

**Comment by snigdha-px** [comment]
> I think your test is not as per the steps in testrail. Testrail steps are as follows:
> 
> ```
> 1. Take a set of backups and share it with multiple users.
> 2. Remove the backup location after sharing it with all users.
> 3. Add it back to the cluster and check from other users if they show up.
> 4. Reshare backups and see if they are successful or not.
> ```
> 
> But in your automated test steps 3 and 4 are missing. You are taking the backup again which is not needed. We have to delete the backup location and add it again. Then re-share it with all users and check the access levels.

Steps are discussed with @kshithijiyer-px and then automated. Will have to change/add  in the testrail also as per what is automated

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:5601]: done!!

**Comment by mkoppal-px** [comment]
@kshithijiyer-px Please check the steps and validation. Waiting for your approval.

**Review by mkoppal-px — APPROVED** [review]
LGTM 👍

**Comment by kshithijiyer-px** [comment]
The steps are correct we have just added one more step of taking a new backup instead of waiting for sync to happen. LGTM

**Review by kshithijiyer-px — APPROVED** [review]

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:2948]: done!!

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:2988]: There are 2 `if err != nil` here, just keep the one that returns more verbose error

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:2863]: If you don't want to return verbose error you can just do
```
return CreateS3BackupLocation(name, uid, cloudCred, cloudCredUID, bucketName, orgID, encryptionKey)
```

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:2764]: just return error, no need to define it
```
return fmt.Errorf("unable to fetching px-central-admin ctx: %v", err)
```

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:2769]: Same here
```
return fmt.Errorf("failed to delete backup location with request: %v", err)
```

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:5521]: Shouldn't this func also return some errors if it failed?

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:3754]: you can just return, since you not doing anything else with the `err` you getting
```
CreateAzureBackupLocation(name, uid, cloudCred, cloudCredUID, bucketName, orgID)
```

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:3750]: Why do you have this function, that all it does is calls the actual function `CreateAzureBackupLocation()` that does the actual backup creating on Azure, this seems redundant, can we remove it and just call the `CreateAzureBackupLocation()` instead where you need to?

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:2860]: Same here, why do you need to create another function instead of just using `CreateS3BackupLocation()`, can we remove it?

**Review by nikolaypopov — CHANGES_REQUESTED** [review]

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:2863]: done

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:2988]: Added by mistake. Now corrected it

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:2764]: done

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:2769]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:5521]: Yes Nikolay You are correct .
These are the changes we are addressing in future PRs

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:3754]: done

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:3750]: Yes Nikolay these are the things we will be addressing in future PRs.

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:2860]: Sure will take this in another PR

File: tests/common.go
Comment (nikolaypopov) [reply tests/common.go:3745]: You are touching this whole function in this PR, why not just remove it and use `CreateS3BackupLocation()` instead?

Comment (snigdha-px) [reply tests/common.go:3745]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:5521]: done

**Review by lsrinivas-pure — APPROVED** [review]
LGTM

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:6004]: nit: remove extra spaces

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:4987]: nit: remove extra spaces

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:4985]: nit: remove extra spaces

File: tests/backup/backup_test.go
Comment (nikolaypopov) [inline tests/backup/backup_test.go:5305]: nit: remove extra spaces

File: tests/common.go
Comment (nikolaypopov) [inline tests/common.go:3652]: nit: remove extra spaces

**Review by nikolaypopov — COMMENTED** [review]
Please fix minor things and it should be good

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:6324]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:6281]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:6149]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:4987]: done

File: tests/backup/backup_test.go
Comment (snigdha-px) [reply tests/backup/backup_test.go:4985]: done

File: tests/common.go
Comment (snigdha-px) [reply tests/common.go:3652]: done

**Review by nikolaypopov — APPROVED** [review]
LGTM

**Pull request merged**
